Windows clang build fails with `-Werror,-Wcast-function-type-mismatch`

### Problem:

When building AWS-LC for Windows using clang and mingw-w64, the build fails with:
```
/path/to/awslc/crypto/rand_extra/windows.c:68:22: error: cast from 'FARPROC' (aka 'long long (*)()')
to 'ProcessPrngFunction' (aka 'int (*)(unsigned char *, unsigned long long)') converts
to incompatible function type [-Werror,-Wcast-function-type-mismatch]
   68 |   g_processprng_fn = (ProcessPrngFunction)GetProcAddress(hmod, "ProcessPrng");
      |                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```
Full log: https://github.com/curl/curl-for-win/actions/runs/11712448493/job/32623148501#step:3:3996

#### Relevant details
AWS-LC commit: 1.37.0

System information: for linux, below info can be collected by running `uname -srvmp`
 * CPU architecture: x86_64
 * OS: Windows (cross-built from Debian testing)

Build log: 
```text
-- The C compiler identification is Clang 19.1.2
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/clang-19 - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- The CXX compiler identification is Clang 19.1.2
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/clang++-19 - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Go not found. Disabling some code generation and using pre-generated code in generated-src/
-- Perl not found. Disabling some code generation and using pre-generated code in generated-src/
-- Setting CMAKE_C_STANDARD=11
-- stdalign_check.c probe is positive, enabling AWS_LC_STDALIGN_AVAILABLE
-- builtin_swap_check.c probe is positive, enabling AWS_LC_BUILTIN_SWAP_SUPPORTED
-- The ASM_NASM compiler identification is NASM
-- Found assembler: /usr/bin/nasm
Copying platform assembly files from /home/runner/work/curl-for-win/curl-for-win/awslc/generated-src/win-x86_64/crypto/ to /home/runner/work/curl-for-win/curl-for-win/awslc/_x64-win-ucrt-bld/crypto
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Configuring done (1.1s)
-- Generating done (0.1s)
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -404,8 +404,10 @@ if(GCC OR CLANG)
   endif()
   set(C_CXX_FLAGS "${C_CXX_FLAGS} -Werror -Wformat=2 -Wsign-compare -Wmissing-field-initializers -Wwrite-strings")
 
-  if(GCC AND CMAKE_C_COMPILER_VERSION VERSION_GREATER "8")
+  if((GCC AND CMAKE_C_COMPILER_VERSION VERSION_GREATER "8") OR
+     (CMAKE_C_COMPILER_ID MATCHES "Clang" AND NOT CMAKE_C_COMPILER_VERSION VERSION_LESS "13"))
     # GCC 8.x added a warning called -Wcast-function-type to the -Wextra umbrella.
+    # Also suppress for all clang versions supporting this warning.
     set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wno-cast-function-type")
   endif()
 

--- a/crypto/rand_extra/windows.c
+++ b/crypto/rand_extra/windows.c
@@ -65,7 +65,7 @@ static void init_processprng(void) {
   if (hmod == NULL) {
     abort();
   }
-  g_processprng_fn = (ProcessPrngFunction)GetProcAddress(hmod, "ProcessPrng");
+  g_processprng_fn = (ProcessPrngFunction)(void(*)(void))GetProcAddress(hmod, "ProcessPrng");
   if (g_processprng_fn == NULL) {
     abort();
   }

